src/playlist.rs

```diff
@@ -354,34 +356,61 @@ impl SessionKey {
     }
 }
 
+#[derive(Debug, PartialEq, Clone)]
+pub enum SessionDataField {
+    Value(String),
+    Uri(String)
+}
+
 /// [`#EXT-X-SESSION-DATA:<attribute-list>`]
 /// (https://tools.ietf.org/html/draft-pantos-http-live-streaming-19#section-4.3.4.4)
-/// The EXT-X-SESSION-KEY tag allows encryption keys from Media Playlists
-/// to be specified in a Master Playlist.  This allows the client to
-/// preload these keys without having to read the Media Playlist(s) first.
-#[derive(Debug, Default, PartialEq, Clone)]
+/// The EXT-X-SESSION-DATA tag allows arbitrary session data to be carried
+/// in a Master Playlist.
+#[derive(Debug, PartialEq, Clone)]
 pub struct SessionData {
     pub data_id: String,
-    pub value: String,
-    pub uri: String,
+    pub field: SessionDataField,
     pub language: Option<String>,
 }
 
 impl SessionData {
-    pub fn from_hashmap(mut attrs: HashMap<String, String>) -> SessionData {
-        SessionData {
-            data_id: attrs.remove("DATA-ID").unwrap_or_else(String::new),
-            value: attrs.remove("VALUE").unwrap_or_else(String::new),
-            uri: attrs.remove("URI").unwrap_or_else(String::new),
-            language: attrs.remove("LANGUAGE"),
+    pub fn from_hashmap(mut attrs: HashMap<String, String>) -> Result<SessionData, String> {
+        let data_id = match attrs.remove("DATA-ID") {
+            Some(data_id) => data_id,
+            None => return Err("EXT-X-SESSION-DATA field without DATA-ID".to_string())
+        };
+
+        let value = attrs.remove("VALUE");
+        let uri = attrs.remove("URI");
+
+        let field = if let Some(value) = value {
+            if uri.is_some() {
+                return Err(format!["EXT-X-SESSION-DATA tag {} contains both a value and a uri", data_id]);
+            }
+            SessionDataField::Value(value)
         }
+        else if let Some(uri) = uri {
+            SessionDataField::Uri(uri)
+        } else {
+          return Err(format!["EXT-X-SESSION-DATA tag {} must contain either a value or a uri", data_id]);
+        };
```

Comment:
Nice! Although one thing I think is somewhat an antipattern is relying too much on `if let Some(thing)/.is_some()`. Instead, pattern matching on both should be possible here:
```
let field = match (value, uri) {
    (Some(_), Some(_)) => return ...,
    (Some(v), None) => SessionDataField::Value(v),
   //...etc
};
```

Reply:
Much neater :) Fixed.